fvm/evm/offchain/utils/collection_test.go

```diff
@@ -2,45 +2,228 @@ package utils_test
+
+	// replay each event files
+	for _, eventsFilePath := range eventsFilePaths {
+		log.Info().Msgf("replaying events from %v, nextHeight: %v", eventsFilePath, nextHeight)
+
+		checkpointEndHeight := replayEvents(t, chainID, store, eventsFilePath, checkpointDir, nextHeight)
+		nextHeight = checkpointEndHeight + 1
+	}
+
+	log.Info().
+		Msgf("succhessfully replayed all events and state changes are consistent with onchain state change. nextHeight: %v", nextHeight)
+}
 
+func initStorageWithCheckpoints(t *testing.T, chainID flow.ChainID, checkpointDir string, checkpointEndHeight uint64) (
+	*TestValueStore, uint64,
+) {
 	rootAddr := evm.StorageAccountAddress(chainID)
 
-	// setup the rootAddress account
-	as := environment.NewAccountStatus()
-	err := storage.SetValue(rootAddr[:], []byte(flow.AccountStatusKey), as.ToBytes())
+	// if there is no checkpoint, create a empty store and initialize the account status,
+	// return 0 as the genesis height
+	if checkpointEndHeight == 0 {
+		store := GetSimpleValueStore()
+		as := environment.NewAccountStatus()
+		require.NoError(t, store.SetValue(rootAddr[:], []byte(flow.AccountStatusKey), as.ToBytes()))
+
+		return store, 0
+	}
+
+	valueFileName, allocatorFileName := checkpointFileNamesByEndHeight(checkpointDir, checkpointEndHeight)
+	values, err := deserialize(valueFileName)
 	require.NoError(t, err)
+	allocators, err := deserializeAllocator(allocatorFileName)
+	require.NoError(t, err)
+	store := GetSimpleValueStorePopulated(values, allocators)
+	return store, checkpointEndHeight
+}
 
-	bp, err := blocks.NewBasicProvider(chainID, storage, rootAddr)
+func replayEvents(
+	t *testing.T,
+	chainID flow.ChainID,
+	store *TestValueStore, eventsFilePath string, checkpointDir string, initialNextHeight uint64) uint64 {
+
+	rootAddr := evm.StorageAccountAddress(chainID)
+
+	bpStorage := storage.NewEphemeralStorage(store)
+	bp, err := blocks.NewBasicProvider(chainID, bpStorage, rootAddr)
 	require.NoError(t, err)
 
+	nextHeight := initialNextHeight
+
+	scanEventFilesAndRun(t, eventsFilePath,
+		func(blockEventPayload *events.BlockEventPayload, txEvents []events.TransactionEventPayload) error {
+			if blockEventPayload.Height != nextHeight {
+				return fmt.Errorf(
+					"expected height for next block event to be %v, but got %v",
+					nextHeight, blockEventPayload.Height)
+			}
+
+			err = bp.OnBlockReceived(blockEventPayload)
+			require.NoError(t, err)
+
+			sp := NewTestStorageProvider(store, blockEventPayload.Height)
+			cr := sync.NewReplayer(chainID, rootAddr, sp, bp, zerolog.Logger{}, nil, true)
+			res, err := cr.ReplayBlock(txEvents, blockEventPayload)
+			require.NoError(t, err)
+
+			// commit all changes
+			for k, v := range res.StorageRegisterUpdates() {
+				err = store.SetValue([]byte(k.Owner), []byte(k.Key), v)
+				require.NoError(t, err)
+			}
+
+			err = bp.OnBlockExecuted(blockEventPayload.Height, res)
+			require.NoError(t, err)
+
+			// commit all block hash list changes
```

Comment:
The `ReplayBlock` method will verify the state change checksum, however, the register updates for the block hash list can not be verified with the block events, because the events does not include the block hash list changes. 

The block hash list changes has to be verified with a separate test case.